shared/src/main/scala/fs2/interop/cats/IOAsyncInstances.scala

```diff
@@ -0,0 +1,169 @@
+          if (id == nonce) {
+            nonce += 1L; val id2 = nonce
+            waiting.values.foreach(cb => ec { cb(r.right.map((_,id2))) })
+            waiting = LinkedMap.empty
+            result = r
+            cb(Right(true))
+          }
+          else cb(Right(false))
+
+        case Msg.Nevermind(id, cb) =>
+          val interrupted = waiting.get(id).isDefined
+          waiting = waiting - id
+          val _ = ec { cb (Right(interrupted)) }
+      }
+
+      new Ref(actor)
+    }
+
+    class Ref[A] private[fs2](actor: Actor[Msg[A]])(implicit ec: ExecutionContext, protected val F: Async[IO]) extends Async.Ref[IO,A] {
+
+      def access: IO[(A, Either[Throwable,A] => IO[Boolean])] =
+        IO(new MsgId {}).flatMap { mid =>
+          getStamped(mid).map { case (a, id) =>
+            val set = (a: Either[Throwable,A]) =>
+              IO.async[Boolean] { cb => actor ! Msg.TrySet(id, a, cb) }
+            (a, set)
+          }
+        }
+
+      /**
+       * Return a `IO` that submits `t` to this ref for evaluation.
+       * When it completes it overwrites any previously `put` value.
+       */
+      def set(t: IO[A]): IO[Unit] =
+        IO { ec { t.unsafeRunAsync { r => actor ! Msg.Set(r) } }; () }
+      def setFree(t: Free[IO,A]): IO[Unit] =
+        set(t.run(F))
+      def runSet(e: Either[Throwable,A]): Unit =
+        actor ! Msg.Set(e)
+
+      private def getStamped(msg: MsgId): IO[(A,Long)] =
+        IO.async[(A,Long)] { cb => actor ! Msg.Read(cb, msg) }
+
+      /** Return the most recently completed `set`, or block until a `set` value is available. */
+      override def get: IO[A] = IO(new MsgId {}).flatMap { mid => getStamped(mid).map(_._1) }
+
+      /** Like `get`, but returns a `IO[Unit]` that can be used cancel the subscription. */
+      def cancellableGet: IO[(IO[A], IO[Unit])] = IO {
+        val id = new MsgId {}
+        val get = getStamped(id).map(_._1)
+        val cancel = IO.async[Unit] {
+          cb => actor ! Msg.Nevermind(id, r => cb(r.right.map(_ => ())))
+        }
+        (get, cancel)
+      }
+
+      /**
+       * Runs `t1` and `t2` simultaneously, but only the winner gets to
+       * `set` to this `ref`. The loser continues running but its reference
+       * to this ref is severed, allowing this ref to be garbage collected
+       * if it is no longer referenced by anyone other than the loser.
+       */
+      def setRace(t1: IO[A], t2: IO[A]): IO[Unit] = IO {
+        val ref = new AtomicReference(actor)
+        val won = new AtomicBoolean(false)
+        val win = (res: Either[Throwable, A]) => {
+          // important for GC: we don't reference this ref
+          // or the actor directly, and the winner destroys any
+          // references behind it!
+          if (won.compareAndSet(false, true)) {
+            val actor = ref.get
+            ref.set(null)
+            actor ! Msg.Set(res)
+          }
+        }
+        t1.unsafeRunAsync(win)
```

Comment:
Hmm.  You're right that they should.  What concerns me though is this code is copy/pasted from fs2-scalaz, where it was written against `scalaz.Task` which would similarly require thread-shifting.